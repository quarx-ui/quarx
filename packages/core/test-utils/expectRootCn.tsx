import React, { FC } from 'react';
import { render } from '@testing-library/react';

export const expectRootCn = <T extends FC<{ className?: string, children: any }>>(Component: T, name: string) => {
    const className = 'test-class-name';
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const { container: { firstChild: rootNode } } = render(<Component className={className} />);
    expect(rootNode).toHaveClass(className);
    expect(rootNode).toHaveClass(name);
};
