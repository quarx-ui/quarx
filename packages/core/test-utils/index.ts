export { expectRootCn } from './expectRootCn';
export * from './testBEM';
