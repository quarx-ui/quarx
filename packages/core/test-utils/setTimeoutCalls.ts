interface FakeSetTimeout {
    mock: {
        calls: Array<[() => void]>; // TODO Доработать типизацию
    };
}

/** При рендере jest использует setTimeout(_flushCallback, 0),
 * из-за чего проверка количества вызовов setTimeout происходит нестабильно.
 *
 * @return массив вызовов setTimeout без _flushCallback */
export const setTimeoutCalls = () => (setTimeout as unknown as FakeSetTimeout)
    .mock.calls
    .filter(([fn]) => !String(fn).includes('_flushCallback'));
