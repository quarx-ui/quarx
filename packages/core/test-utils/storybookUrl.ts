type PropsType = Record<string, string | boolean | number>;

interface UrlOptions {
    component: string;
    story?: string;
    props?: PropsType;
}

const reducePropsToURLString = (props?: PropsType) => {
    if (!props) {
        return '';
    }

    const keysOfProps = Object.keys(props);
    const valuesOfProps = Object.values(props);

    if (keysOfProps.length > 1) {
        return keysOfProps.reduce((prev, cur) => {
            const currentProp = `${cur}:${props[cur]}`;
            return `${prev};${currentProp}`;
        });
    }
    return `${keysOfProps[0]}:${valuesOfProps[0]}`;
};

export function storyUrl(options: UrlOptions) {
    const { component, story = 'sandbox', props } = options;
    const reducedProps = reducePropsToURLString(props);

    return `?path=/story/core-${component}--${story}&args=${reducedProps}`;
}
