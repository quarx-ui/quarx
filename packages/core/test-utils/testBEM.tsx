import React, { ComponentType } from 'react';
import { screen, render, cleanup } from '@testing-library/react';

export const expectPropsInClasses = (
    element: HTMLElement,
    qxClassname?: string,
) => (
    prop: string,
    value?: string | boolean,
) => {
    const qxSelector = qxClassname ?? element.classList[0];

    const valueIsBoolean = typeof value === 'boolean' || value === 'true' || value === 'false';
    const booleanClassname = value === true || value === 'true' ? `${qxSelector}_${prop}` : undefined;

    const expectedClassname = booleanClassname
        ?? (!valueIsBoolean
            ? `${qxSelector}_${prop}_${value}`
            : undefined);

    if (!expectedClassname) {
        return null;
    }

    return expect(element).toHaveClass(expectedClassname);
};

export const expectPropsMapInClasses = (element: HTMLElement, qxClassname?: string) => {
    const expectPropInClasses = expectPropsInClasses(element, qxClassname);

    return (expectedProps: Record<string, string | boolean>) => {
        Object.keys(expectedProps).forEach((prop) => {
            expectPropInClasses(prop, expectedProps[prop]);
        });
    };
};

export const testStyleParams = <
    S extends Record<string, any>,
    P extends Record<string, any> = Record<string, any>,
>(
    Component: ComponentType<P>,
    expectedDefaultParams: S,
    defaultProps: P = {} as P,
    qxClassname?: string,
) => (props: { [key in keyof S]: S[key][] }) => {
    describe('style params', () => {
        describe('default values', () => {
            const testId = 'testStyleParams-default';

            render(
                <Component
                    {...defaultProps}
                    data-testid={testId}
                />,
            );

            const element = screen.getByTestId(testId);

            Object.entries(expectedDefaultParams).forEach(([key, value]) => {
                it(`${key}`, () => {
                    expectPropsInClasses(element, qxClassname)(key, value);
                });
            });

            cleanup();
        });

        Object.entries(props).forEach(([propName, values]) => {
            describe(propName, () => {
                Object.values(values).forEach((value) => {
                    it(`${value}`, () => {
                        const testId = `testStyleParams-${propName}-${value}`;

                        const testProps = { [propName]: value };
                        const { asFragment } = render(
                            <Component
                                {...(defaultProps ?? {})}
                                {...testProps as P}
                                data-testid={testId}
                            />,
                        );

                        const element = screen.getByTestId(testId);
                        expectPropsMapInClasses(element, qxClassname)(testProps as S);
                        expect(asFragment()).toMatchSnapshot();
                    });
                });
            });
        });
    });
};
