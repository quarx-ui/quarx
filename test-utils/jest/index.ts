export { testRootCn } from './testRootCn';
export * from './testBEM';
