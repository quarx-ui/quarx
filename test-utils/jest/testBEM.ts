export const expectPropsInClasses = (element: HTMLElement, sxClassname?: string) => (prop: string, value?: string | boolean) => {
    const sxSelector = sxClassname ?? element.classList[0];

    const valueIsBoolean = typeof value === 'boolean' || value === 'true' || value === 'false';
    const booleanClassname = value === true || value === 'true' ? `${sxSelector}_${prop}` : undefined;

    const expectedClassname = booleanClassname
        ?? (!valueIsBoolean
            ? `${sxSelector}_${prop}_${value}`
            : undefined);

    if (!expectedClassname) {
        return null;
    }

    return expect(element).toHaveClass(expectedClassname);
};

export const expectPropsMapInClasses = (element: HTMLElement, sxClassname?: string) => {
    const expectPropInClasses = expectPropsInClasses(element, sxClassname);

    return (expectedProps: Record<string, string | boolean>) => {
        Object.keys(expectedProps).forEach((prop) => {
            expectPropInClasses(prop, expectedProps[prop]);
        });
    };
};
