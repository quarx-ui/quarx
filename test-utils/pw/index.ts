export {
    compareStorySnapshots,
    compareVariantSnapshots,
    comparePropertySnapshots,
    comparePropertiesMapSnapshots,
} from './compareSnapshots';
