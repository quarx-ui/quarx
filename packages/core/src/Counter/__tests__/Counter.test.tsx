import { render } from '@testing-library/react';
import React from 'react';
import { Counter } from '@core';
import { testRootCn } from '../../../../../test-utils/jest';

describe('Counter', () => {
    it('Snapshot', () => {
        const { asFragment } = render(
            <Counter>
                99
            </Counter>,
        );
        expect(asFragment())
            .toMatchSnapshot();
    });

    it('className', () => {
        testRootCn(Counter, 'SxCounter');
    });
});
