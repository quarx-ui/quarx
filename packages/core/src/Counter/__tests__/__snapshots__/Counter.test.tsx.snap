// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Snapshot 1`] = `
<DocumentFragment>
  <span
    class="SxCounter SxCounter_type_filled SxCounter_size_large SxCounter_color_brand css-9229zm-SxCounter-root"
  >
    99
  </span>
</DocumentFragment>
`;
